core/src/main/java/org/elasticsearch/cluster/service/InternalClusterService.java

```diff
@@ -179,6 +150,11 @@ public void setClusterStatePublisher(BiConsumer<ClusterChangedEvent, Discovery.A
         clusterStatePublisher = publisher;
     }
 
+    public void setLocalNode(DiscoveryNode localNode) {
+        DiscoveryNodes.Builder nodeBuilder = DiscoveryNodes.builder(clusterState.nodes()).put(localNode).localNodeId(localNode.id());
```

Comment:
we have to have some assertion that this is executed before we are started. And it should also be synchronized as well as doStart()